swarm/README.md

```diff
@@ -0,0 +1,53 @@
+Mini-broker Swarm deployment for isolated containerized filters
+--------------------------------------------------------------
+
+These are Swarm deployment scripts for a version of the
+mini-broker running a one-container-per-filter setup.
+This design is updated from the design described in DMTN-081
+[https://dmtn-028.lsst.io](https://dmtn-028.lsst.io), which
+runs multiple filters in the same container.
+To deploy a prototype of the full mini-broker with 100 consumers
+(end users) and 100 individual filters each in their own container,
+run the `deploy_all.sh` script.
+This will deploy the following setup on a total of 12 nodes (one manager and
+11 worker nodes):
+
+* A main Kafka/Zookeeper hub on the Swarm manager node that receives the full alert stream and runs by itself on the manager node
+* A single alert producer running on its own separate node that sends the full stream of alerts to the main Kafka/Zookeeper hub
+* Five downstream Kafka/Zookeepers for feeding the filterers, which mirror the full stream and run each on a separate node
+* 100 stream filterer containers split into 20 on each of the five Kafka/Zookeeper mirror nodes, which read the full stream and send filtered alert streams back to the co-located Kafka mirror
+* 100 consumers of the filtered streams running as 20 separate containers on five separate nodes and read from the Kafka mirrors
+
+Required are `consumernodes.txt`, `filternodes.txt`, and `sendernodes.txt`,
+each with the Swarm IDs of the nodes on which to run the different components
+on separate lines.  
+The consumer and filter nodes files should each have the IDs of five unique nodes,
+so that each node will run 20 consumer or 20 filter containers, for a total
+of 100 consumers and 100 filters.
+The sender node file needs one Swarm ID on which to place a single
+alert producer sending 10,000 alerts per visit (every 39 seconds).
+
+Performance
+-----------
+
+This design has the advantage of being able to run 100 filters without
+each filter having to read its own full copy of the stream over a network.
+With groups of 20 filters on five Kafka mirror nodes, the network  
+needs only to support enough bandwidth for the alert producer to the
+main Kafka hub and five full streams feeding each of the mirror nodes.
+Individual filters read from and write to co-located Kafka mirrors.
```

Comment:
A mode detailed description of the distribution of the 100 filters, and the interaction between nodes might be helpful. Is each filter self contained, or do they talk to each other in some fashion? Does each Kafka node get a full copy of the stream? If a filter does not get a full copy of the stream, is that because the stream was prefiltered, or is it some kind of subset that you have worked out in advance?

Reply:
Are there disadvantages to this design? Is it less flexible than 100 separate streams, or require more resources of some other sort (memory?), or slow the rest of the filters on a node to the speed of the slowest individual filter?

Reply:
Added some comments here.